optional/slides/slides.md

```diff
@@ -0,0 +1,207 @@
+Conversely, the absence of a value is modeled with an “empty” optional returned by the method `Optional.empty()`.
+
+You might wonder what the difference is between a null reference and `Optional.empty()` which is a valid, workable object of type Optional that can be invoked in useful ways.
+
+???
+Absent is a static factory method that returns a special singleton instance of the Optional class.
+
+---
+
+## Introducing the Optional class, for Java 6
+
+### Yes! Guava has `com.google.common.base.Optional<T>` too!
+
+Unlike the Java one this implements _Serializable_!
+
+Has similar methods but with different names:
+* `fromNullable` and `ofNullable`
+* `absent` and `empty`
+* `or` and `orElse`
+* `transform` and `map`/`flatMap`
+
+But sometimes not: `get`, `isPresent`.
+
+???
+This can be used like a field.
+
+---
+
+## Example: Guava
+
+```
+public class Insurance {
+    private String name;
+    public String getName() { return name; }
+}
+
+public String getInsuranceName(Insurance insurance) {
+	return Optional.fromNullable(insurance)
+		.transform(new Function<Insurance, String>() {
+			@Override public String apply(Insurance input) {
+				return input.getName();
+			}
+    }).or("Unknow");
+}
+```
+
+Java 6 style! Still ugly...
+
+---
+
+## Example: Method returns Optional&lt;T&gt;
+
+Returning Optional adds explicitly the possibility that there might not be a name for that given Insurance.
+
+This means that the caller of the method is explicitly forced by the type system to think about and deal with the possibility that there might not be a name with that Insurance.
+
+```
+public class Insurance {
+    private String name;
+    public Optional<String> getName() {
+    	return Optional.fromNullable(name);
+    }
+}
+```
+
+---
+
+## Example: Java 8
+
+```
+public class Insurance {
+	private String name;
+	public Optional<String> getName() {
+		return Optional.ofNullable(name);
+	}
+}
+```
+```
+insurance.getName().orElse("Unknown");
+```
+
+Bye Bye NullPointerException!
+
+---
+
+## Interesting methods
+
+### Guava
+
+* `Set<T> asSet()`: immutable singleton Set if it is present; an empty immutable Set otherwise.
+* `T or(T)`: contained instance if it is present; defaultValue otherwise.
+* `Optional<V> transform(Function<? super T, V>)`: if the instance is present, it is transformed with the given Function; otherwise, Optional.absent is returned.
+
+---
+
+## Interesting methods
+
+### Java
+
+* `Optional<T> filter(Predicate<? super T>)`: when predicate is false return an empty optional
+* `Optional<U> map(Function<? super T, ? extends U>)`: if a value is present, apply the provided mapping function to it
+* `Optional<U> flatMap(Function<? super T, Optional<U>>)`: like map, but the function result in another optional
+* `void ifPresent(Consumer<? super T>)`: if a value is present, invoke the specified consumer with the value
+* `T orElse(T)`
+* `<X extends Throwable> T orElseThrow(Supplier<? extends X>)`
```

Comment:
these two are available also in guava, and, imho, are **the** most useful methods of the optional.